staging/src/k8s.io/client-go/tools/cache/controller_test.go

```diff
@@ -893,3 +896,224 @@ func (m *mockTxnStore) Transaction(txns ...Transaction) *TransactionError {
+			return processDeltasInBatch(recorder, store, deltaList, isInInitialList)
+		},
+	}
+
+	c := New(cfg)
+	go c.RunWithContext(ctx)
+	if !WaitForCacheSync(ctx.Done(), c.HasSynced) {
+		t.Fatal("Timed out waiting for cache sync")
+	}
+	testReplaceEvents(t, ctx, fifo, recorder, store)
+}
+
+func testReplaceEvents(t *testing.T, ctx context.Context, fifo Queue, m *eventRecorder, store Store) {
+	tcs := []struct {
+		name            string
+		initialObjs     []metav1.Object
+		replacedItems   []interface{}
+		expectedHistory []eventRecord
+	}{
+		{
+			name: "Create with no initial objects",
+			replacedItems: []interface{}{
+				&v1.Pod{ObjectMeta: metav1.ObjectMeta{Name: "pod-1", Namespace: "default"}},
+			},
+			expectedHistory: []eventRecord{
+				{Action: "add", Key: "default/pod-1"},
+			},
+		},
+		{
+			name: "Delete all objects",
+			initialObjs: []metav1.Object{
+				&v1.Pod{ObjectMeta: metav1.ObjectMeta{Name: "pod-1", Namespace: "default"}},
+				&v1.Pod{ObjectMeta: metav1.ObjectMeta{Name: "pod-3", Namespace: "default"}},
+			},
+			replacedItems: []interface{}{},
+			expectedHistory: []eventRecord{
+				{Action: "delete", Key: "default/pod-1"},
+				{Action: "delete", Key: "default/pod-3"},
+			},
+		},
+		{
+			name: "Create, update, delete",
+			initialObjs: []metav1.Object{
+				&v1.Pod{ObjectMeta: metav1.ObjectMeta{Name: "pod-1", Namespace: "default"}},
+				&v1.Pod{ObjectMeta: metav1.ObjectMeta{Name: "pod-2", Namespace: "default", Labels: map[string]string{"ver": "1"}}},
+				&v1.Pod{ObjectMeta: metav1.ObjectMeta{Name: "pod-3", Namespace: "default"}},
+			},
+			replacedItems: []interface{}{
+				&v1.Pod{ObjectMeta: metav1.ObjectMeta{Name: "pod-1", Namespace: "default"}},
+				&v1.Pod{ObjectMeta: metav1.ObjectMeta{Name: "pod-2", Namespace: "default", Labels: map[string]string{"ver": "2"}}},
+				&v1.Pod{ObjectMeta: metav1.ObjectMeta{Name: "pod-4", Namespace: "default"}},
+			},
+			expectedHistory: []eventRecord{
+				{Action: "update", Key: "default/pod-1"},
+				{Action: "update", Key: "default/pod-2"},
+				{Action: "delete", Key: "default/pod-3"},
+				{Action: "add", Key: "default/pod-4"},
+			},
+		},
+	}
+
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			for _, obj := range tc.initialObjs {
+				require.NoError(t, fifo.Add(obj), "failed to add")
+			}
+
+			// Wait for all create events to appear
+			err := m.waitForEventCount(ctx, len(tc.initialObjs), 5*time.Second)
+			if err != nil {
+				t.Fatalf("Controller failed to receive all initial setup events, got: %v", m.getHistory())
+			}
+
+			// Clear history so we only compare what happens during the RESYNC/REPLACE
+			m.clearHistory()
+
+			// Run replace
+			require.NoError(t, fifo.Replace(tc.replacedItems, "123"), "failed to replace")
+
+			err = m.waitForEventCount(ctx, len(tc.expectedHistory), 5*time.Second)
+			if err != nil {
+				t.Fatalf("Controller failed to receive all events, got: %v", m.getHistory())
+			}
```

Comment:
require.NoError supports string formatting.


```suggestion
			require.NoError(t, m.waitForEventCount(ctx, len(tc.expectedHistory), 5*time.Second), "Controller failed to receive all events, got: %v", m.getHistory())
```

Reply:
Moved to use require.NoError